src/Features/Core/Portable/Extensions/ExtensionMessageHandlerService.cs

```diff
@@ -0,0 +1,358 @@
+
+                    if (duplicateHandler is not null)
+                    {
+                        assemblyHandlers = null;
+                        throw new InvalidOperationException($"Handler name {duplicateHandler} is already registered.");
+                    }
+
+                    foreach (var handler in assemblyHandlers.WorkspaceMessageHandlers)
+                    {
+                        _workspaceHandlers.Add(handler.Key, handler.Value);
+                    }
+
+                    foreach (var handler in assemblyHandlers.DocumentMessageHandlers)
+                    {
+                        _documentHandlers.Add(handler.Key, handler.Value);
+                    }
+                }
+            }
+            finally
+            {
+                extension.SetAssemblyHandlers(assemblyFileName, assemblyHandlers);
+            }
+        }
+    }
+
+    private sealed class Extension(ExtensionMessageHandlerService extensionMessageHandlerService, IAnalyzerAssemblyLoaderInternal analyzerAssemblyLoader, string assemblyFolderPath)
+    {
+        /// <summary>
+        /// Gets the object that is used to lock in order to avoid multiple calls from the same extensions to load the same assembly concurrently
+        /// resulting in the constructors of the same handlers being called more than once.
+        /// All other concurrent operations, including modifying <see cref="_assemblies"/> are protected by
+        /// <see cref="ExtensionMessageHandlerService._lockObject"/>.
+        /// </summary>
+        private readonly object _assemblyLoadLockObject = new();
+
+        private readonly Dictionary<string, AssemblyHandlers?> _assemblies = new();
+
+        private readonly ExtensionMessageHandlerService _extensionMessageHandlerService = extensionMessageHandlerService;
+
+        public IAnalyzerAssemblyLoaderInternal AnalyzerAssemblyLoader { get; } = analyzerAssemblyLoader;
+
+        public string AssemblyFolderPath { get; } = assemblyFolderPath;
+
+        public void SetAssemblyHandlers(string assemblyFileName, AssemblyHandlers? value)
+        {
+            EnsureGlobalLockIsOwned();
+            _assemblies[assemblyFileName] = value;
+        }
+
+        public bool TryGetAssemblyHandlers(string assemblyFileName, out AssemblyHandlers? value)
+        {
+            EnsureGlobalLockIsOwned();
+            return _assemblies.TryGetValue(assemblyFileName, out value);
+        }
+
+        public bool RemoveAssemblyHandlers(string assemblyFileName, out AssemblyHandlers? value)
+        {
+            EnsureGlobalLockIsOwned();
+            return _assemblies.Remove(assemblyFileName, out value);
+        }
+
+        public int AssemblyHandlersCount
+        {
+            get
+            {
+                EnsureGlobalLockIsOwned();
+                return _assemblies.Count;
+            }
+        }
+
+        public ValueTask<RegisterExtensionResponse> LoadAssemblyAsync(string assemblyFileName)
```

Comment:
nothing looks async here.  make synchronous.  caller can be async if they need/want to be.